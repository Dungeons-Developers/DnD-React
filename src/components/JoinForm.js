import React, {useEffect} from 'react';
import {connect} from 'react-redux';
import {If} from 'react-if';
import {Redirect} from 'react-router-dom';


import {TextField, Container, Button, Paper} from '@material-ui/core';

import useForm from '../hooks/useForm';

import {findCampaign, getUserCampaigns} from '../store/slices/campaign-slice';



function JoinForm({findCampaign, campaignID, campaigns, username, getUserCampaigns}) {

  useEffect(() => {
    getUserCampaigns(username);
  }, [getUserCampaigns, username])

  const defaults = {
    campaignID: ''
  }

  function submit(e) {
    e.preventDefault();
    handleSubmit(findCampaign);
    e.target.reset();
  }

  function formChange(e) {
    handleChange(e.target.id, e.target.value)
  }

  const {handleChange, handleSubmit, fields} = useForm(defaults);

  const styles = {
    container: {
      display: 'flex',
      alignItems: 'center',
      justifyContent: 'center',
      flexDirection: 'column'
    },
    form: {
      display: 'flex',
      alignItems: 'center',
      justifyContent: 'center',
      flexDirection: 'column'
    },
    campList: {
      width: '100%',
      display: 'flex',
      flexWrap: 'wrap',
      justifyContent: 'space-around',
      alignItems: 'center'
    },
    paper: {
      width: '30%',
      padding: '5px',
      marginBottom: '10px'
    }
  }

  return (
    <Container style={styles.container}>
      <div>
        <h1>Enter the ID of the campaign you wish to join.</h1>
        <p>If you don't have any campaigns of your own, create one, or ask a friend to join theirs!</p>
      </div>
      <form autoComplete='off' onSubmit={submit} style={styles.form}>
        <div>
          <TextField 
            id="campaignID" 
            label="CampaignID" 
            color="secondary"
            margin="normal" 
            onChange={formChange} 
            value={fields.campaignID}
          />
        </div>
        <Button 
          variant="contained" 
          color="secondary" 
          type="submit" 
          margin="normal"
          disabled={!fields.campaignID}
        >
          Join
        </Button>
      </form>
      <If condition={campaigns.length > 0}>
        <div>
          <p>Your campaigns:</p>
          <div style={styles.campList}>
            {campaigns.map((cam, i) => (
              <Paper style={styles.paper}>
                <h4>{cam.title}</h4>
                <p>ID: {cam._id}</p>
              </Paper>
            ))}
          </div>
        </div>
      </If>

      <If condition={!!campaignID}>
          <Redirect to='/play'/>
      </If>
    </Container>
  )
}

const mapDispatchToProps = {findCampaign, getUserCampaigns};

const mapStateToProps = state => {
  return {
    campaignID: state.campaign.campaignID,
    campaigns: state.campaign.allCampaigns,
    username: state.users.username
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(JoinForm);