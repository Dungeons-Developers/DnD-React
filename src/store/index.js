import { configureStore } from '@reduxjs/toolkit';
import { combineReducers } from 'redux';

// import slices here

import users from './slices/user-slice';
import campaign from './slices/campaign-slice';
import characters from './slices/character-slice';
import theme from './slices/theme-slice';

let reducers = combineReducers( { users, campaign, characters, theme } );

let store = configureStore( { reducer: reducers } );

export default store;

