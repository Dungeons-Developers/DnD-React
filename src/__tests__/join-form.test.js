import React from 'react';
import {mount} from 'enzyme';

import {Provider} from 'react-redux';
import store from '../store';

import JoinForm from '../components/JoinForm';

describe('testing join form', () => {
  let component = mount(
    <Provider store={store}>
      <JoinForm />
    </Provider>
  );

  it('contains proper components', () => {

    expect(component.find('.join-form')).toBeDefined();
    expect(component).toMatchSnapshot();
  });
});