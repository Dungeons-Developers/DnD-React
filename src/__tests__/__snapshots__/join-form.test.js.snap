// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing join form contains proper components 1`] = `ReactWrapper {}`;
